egg/zoo/compo_vs_generalization/losses.py

```diff
@@ -120,58 +121,79 @@ def forward(
         _aux_input,
     ):
         batch_size = sender_input.size(0)
-        idxs_to_mask = _aux_input['idxs_to_mask']
+        idxs_to_reveal = _aux_input['idxs_to_reveal']
+        n_revealed = _aux_input['n_revealed']
 
         masked_input = mask_attributes(sender_input,
-                idxs_to_mask,
-                self.n_attributes,
-                self.n_values,
-                remove_masked_data=True)
+                                       idxs_to_reveal,
+                                       self.n_attributes,
+                                       self.n_values)
         masked_output = mask_attributes(receiver_output,
-                idxs_to_mask,
-                self.n_attributes,
-                self.n_values,
-                remove_masked_data=True)
+                                        idxs_to_reveal,
+                                        self.n_attributes,
+                                        self.n_values)
 
         masked_input = masked_input.view(
-                batch_size, -1, self.n_values
-            )
+            batch_size, self.n_attributes, self.n_values
+        )
         masked_output = masked_output.view(
-                batch_size, -1, self.n_values
-            )
+            batch_size, self.n_attributes, self.n_values
+        )
 
         # acc : (batch_size,) tensor containing
         # 1 where the vectors were perfectly
         # reconstructed by the receiver, 0 elsewhere.
         acc = (
-            torch.sum(
-                (
-                    masked_output.argmax(dim=-1) == masked_input.argmax(dim=-1)
-                ).detach(),
-                dim=1,
+            (
+                torch.sum(
+                    (
+                        # this counts all masks as correct predictions
+                        masked_output.argmax(dim=-1) == masked_input.argmax(dim=-1)
+                    ).detach(),
+                    dim=1,
+                )
+                # so we have to compensate by
+                # substracting the number of masks
+                - (self.n_attributes - n_revealed)
             )
```

Comment:
This is very hard to read. There is also a lot of redundancy because you are
1. masking the inputs and predictions
2. Computing the losses/accuracies, etc based on these masked input/output
3. Handling the masked positions **again** in the losses/accuracies

Why not skip step 1 since you need to handle the masked attributes at the end anyway? Compute loss/acc for every attribute, and only apply the masks when you are aggregating the metric. This would save you a lot of lines of code.

How about this:

```python
receiver_input = receiver_input.view(batch_size, self.n_attributes, self.n_values)
receiver_output = receiver_output.view(batch_size, self.n_attributes, self.n_values)
# Compute mask for revealed attributes
revealed_mask = torch.zeros_like(receiver_input)
revealed_mask = revealed_mask.scatter(dim=1, index=idxs_to_reveal, value=1).float()
# Matches for each attributes
matches = (receiver_output.argmax(dim=-1) == receiver_input.argmax(dim=-1)).float()
# Average across attributes (but only count revealed attributes)
acc_or = (matches * revealed_mask).sum(-1) / revealed_mask.sum(-1)
# Exact matches: Count 1 if attribute is predicted correctly OR if it is masked.
acc = torch.all((matches == 1) | (revealed_mask == 0), dim=-1).float()
# Loss for each attribute (you need to flatten the first dimensions and reshape them afterwards)
loss_by_attributes = F.cross_entropy(
    receiver_output.view(-1, n_attributes),
    receiver_input.argmax(-1).view(-1),
    reduction="none",
).view(batch_size, n_attributes)
# Take the mean, but only of the revealed attributes
loss = (losses * revealed_mask).sum(-1) / revealed_mask.sum(-1)
```